### DISCUSSION TITLE
Get type of a query

### DISCUSSION THREAD

**Discussion by vietanhlehuu** [trpc/trpc#1863] [Q&A] [answered]
Hi, thanks for creating trpc, it's really great. I wonder there is a way to get the type of a query? 
I'm using trpc with next.js. Here is an example:

```js
function CommentList({ postId }: { postId: string }) {
  const { data: post } = trpc.useQuery(["comments", postId]);

  return <div>
    {
      data.comments.map(
        (comment) => <Comment key={comment.id} comment={data.comment} />
      )
    }
  </div>
}

function Comment({ comment }: Props) { // <--- NEED THE TYPE OF THIS PROPS
  const { data } = trpc.useQuery(["comment", { id: comment.userId }], {
    initialData: comment
  });

  return <div>...</div>
}
```

I want to pass in an initial data, so it needs the type of the query inside. So, how can I get the type of this query `trpc.useQuery(["comment", { id: comment.userId }]` to use it in Props. Thanks in advanced.

**Comment by KATT** [comment, marked as answer]
https://trpc.io/docs/infer-types

**Reply by bazo** [reply]
whatever i try, even the examples from this link, i always get a type like this 
```ts
SerializeObject<UndefinedToOptional<Project>>[]
```

i would like to get just Project[]

how to achieve that? thanks

**Reply by KATT** [reply]
We improved it a bit in one of the latest versions but if you want it non-serialized you can use a data transformer: https://trpc.io/docs/data-transformers

**Reply by bazo** [reply]
the actual data is not serialized, the response contains plain js objects.

just the type says it's serialized...

it's hard to use this type in the rest of the app

btw does anyone know where the SerializeObject type is declared? i couldn't find anything in my node modules or on google :(

**Reply by revero-doug** [reply]
I'm with @bazo ... furthermore I don't think data-transformers apply to this issue/use case at all. As I understand it, one can use zod to validate inputs, and when using the (vanilla or other) client, responses should have the correct type -- it's right on the homepage and it's trpc's main selling point. whether I hover over the response or I use inferRouterOutput, I get types like `SerializeObject<UndefinedToOptional<ActualReturnType>>[]` instead of `ActualReturnType[]`. There's nothing in the docs that addresses this scenario, and this appears to be the only relevant discussion on trpc github.

In case it's helpful, I'll note some things about my setup and additional detail on the issue:

1. I'm using pnpm workspaces for a monorepo and have separate packages for trpc server and trpc client.
2. If I pass the response into an identity function that specifies the correct type for its argument (and implicitly its own return type), typescript does not complain.
3. one of the properties is incorrectly typed as `never` in the response, which would seem to contradict #2, but apparently is generally the case in typescript
4. I'm not using esm, but rather `"esModuleInterop": true` and `"module": "commonjs"`
5. I'm using `"strict": true`

**Reply by Nick-Lucas** [reply]
If you use a transformer this type will go away, its function is to transform a type definition which contains non-json-serialisable data and output a type which matches the result of `JSON.parse(JSON.stringify(myData))` (dates become strings for instance) - this prevents you seeing a date as a date on the UI when in-fact it's a string

If you use a transformer then types like Date will be recognised and deserialised back to dates. tRPC is actually best used with a transformer, but we support without for the rare cases it's not desired

**Reply by revero-doug** [reply]
> If you use a transformer this type will go away, its function is to transform a type definition which contains non-json-serialisable data and output a type which matches the result of `JSON.parse(JSON.stringify(myData))` (dates become strings for instance) - this prevent you seeing a date as a date on the UI when in-fact it's a string
> 
> If you use a transformer then types like Date will be recognised and deserialised back to dates. tRPC is actually best used with a transformer, but we support without for the rare cases it's not desired

@Nick-Lucas ah I suspect `bigint` is causing the issue then. what's the most unobtrusive/minimal way to utilize transformers, given that currently using zod for input validation?

**Reply by Nick-Lucas** [reply]
Zod and SuperJSON (our recommended transformer) don't step on each other's toes, Zod validates and SuperJSON serialised/deserialises. Most tRPC codebases use both together

**Reply by revero-doug** [reply]
@Nick-Lucas tyvm, that was a quick fix! might I recommend adding a short note that includes the telltale `SerializeObject<UndefinedToOptional<<YourType>>` sign and the superjson transformer suggestion as a footnote to the Quick Start example in the docs? this was quite the head scratcher.

**Reply by Nick-Lucas** [reply]
Would you be able to share some thoughts around this here? https://github.com/trpc/trpc/discussions/3963

We have been rewriting a lot of docs, and transformers haven't been touched (yet). Would love your thoughts in full :)

**Comment by jamezening** [comment]
Also much thanks and kudos for `trpc`.

Is it possible to dig out a nested type from the InferQueryOutput?
- for example, if the `Comment[]` type above had a nested `NestedComments[]` type

Or is there another way a nested type should be approached?

**Reply by JonParton** [reply]
You should be able to just use typescript syntax to [index the sub property](https://www.typescriptlang.org/docs/handbook/2/indexed-access-types.html) out of the inferred output type.

A trivialised example as I can't see actual types above to use
```tsx
// Example type... What should have been defined on the server and inferred using the helpers
type Comment = {
     author: User, 
     title: string, 
     nestedComments: NestedComment[]
} 

// Accessing a sub property 
type NestedComments = Comment["nestedComments"] // notice it's the property name ... lower case
// ^ will be of type NestedComment[] 
```

**Reply by jamezening** [reply]
Awesome!

I need to up my TypeScript game - the indexed sub property worked. Thank you!

**Reply by cyotani** [reply]
Thanks for answering this. I'm also new to typescript and was looking for how to do this. It might be worth considering adding this to the "infer-types" docs to make it a little more Noob friendly.

**Reply by mrindzhov** [reply]
Is there an easy way to get the element type only? In this case `NestedComponent`?

**Reply by JonParton** [reply]
You can index it with the generic `number`

```Tsx
type NestedComment = Comment["nestedComments"][number]
```
